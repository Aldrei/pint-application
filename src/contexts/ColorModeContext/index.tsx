import { createContext } from 'react';

/* istanbul ignore next */ 
// eslint-disable-next-line @typescript-eslint/no-empty-function
const ColorModeContext = createContext({ toggleColorMode: () => {} });

export default ColorModeContext;