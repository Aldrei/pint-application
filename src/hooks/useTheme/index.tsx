import React from 'react';

import { createTheme as createThemeOrigin } from '@mui/material/styles';
import { PaletteMode } from '@mui/material';

declare module '@mui/material/styles' {
  interface Theme {
    pallete?: {
      mode: string;
      primary?: {
        main: string;
        dark: string;
      },
      background: {
        default: string;
        paper: string;
      },
      text: {
        primary: string;
        secondary: string;
      },
      divider: string;
    }
  }
  // allow configuration using `createTheme`
  interface ThemeOptions {
    pallete?: {
      mode: string;
      primary?: {
        main: string;
        dark: string;
      },
      background: {
        default: string;
        paper: string;
      },
      text: {
        primary: string;
        secondary: string;
      },
      divider: string;
    }
  }
}

const breakpoints = {
  values: {
    xs: 0,
    sm: 600,
    md: 900,
    lg: 1200,
    xl: 1536,
    mobile: 0,
    tablet: 640,
    laptop: 1024,
    desktop: 1200,
  },
};

const getDesignTokens = (mode: PaletteMode) => ({
  ...breakpoints,
  palette: {
    mode,
    ...(mode === 'light'
      ? {
        text: {
          primary: '#000000',
          secondary: 'rgb(172, 39, 189)'
        },
        background: {
          default: '#de1818',
          paper: '#1f10c6'
        },
        divider: '#c4e715'
      }
      : {
        text: {
          primary: 'rgb(255, 255, 255)',
          secondary: 'rgb(178, 186, 194)'
        },
        background: {
          default: 'rgb(0, 30, 60)',
          paper: 'rgb(10, 25, 41)'
        },
        divider: 'rgba(255, 255, 255, 0.12)'
      }
    ),
  },
});

export const createTheme = (mode: PaletteMode) => createThemeOrigin(getDesignTokens(mode));

const useTheme = () => {
  const [mode, setMode] = React.useState<'light' | 'dark'>('dark');

  const colorMode = React.useMemo(
    () => ({
      toggleColorMode: () => {
        setMode((prevMode) => (prevMode === 'light' ? 'dark' : 'light'));
      },
    }),
    [],
  );

  const theme = React.useMemo(() => createTheme(mode), [mode]);

  return { theme, colorMode };
};

export default useTheme;