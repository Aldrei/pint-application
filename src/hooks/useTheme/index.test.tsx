import * as React from 'react';

import renderThemeProvider from '../../helpers/test/renderThemeProvider';
import ColorModeContext from '../../contexts/ColorModeContext';
import useTheme from '../useTheme';
import { fireEvent } from '@testing-library/react';

const ComponentTest = (): React.ReactElement => {
  const { theme } = useTheme();
  const colorMode = React.useContext(ColorModeContext);

  return (
    <React.Fragment>
      <button onClick={colorMode.toggleColorMode}>Change mode</button>
      <span>{`Theme mode: ${theme.pallete?.mode}`}</span>
    </React.Fragment>
  );
};

describe('useBreakpoints', () => {
  it('Should change to light mode', () => {
    const nodeEl = renderThemeProvider(<ComponentTest />);

    fireEvent.click(nodeEl.getByText('Change mode'));
    expect(nodeEl.baseElement).toMatchSnapshot();

    fireEvent.click(nodeEl.getByText('Change mode'));
    expect(nodeEl.baseElement).toMatchSnapshot();
  });
});
