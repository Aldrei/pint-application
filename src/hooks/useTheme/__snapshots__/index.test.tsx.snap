// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useBreakpoints Should change to light mode 1`] = `
<body>
  <div>
    <button>
      Change mode
    </button>
    <span>
      Theme mode: undefined
    </span>
  </div>
</body>
`;

exports[`useBreakpoints Should change to light mode 2`] = `
<body>
  <div>
    <button>
      Change mode
    </button>
    <span>
      Theme mode: undefined
    </span>
  </div>
</body>
`;
